### ISSUE TITLE
Using BucketSampler makes loading data extremely slow

### ISSUE THREAD

**Issue by Yuancheng-Xu** [a-r-r-o-w/finetrainers#206]
Hi,

When training CogVideoX using cogvideox_text_to_video_lora.py, I found that using `sampler=BucketSampler(train_dataset, batch_size=args.train_batch_size, shuffle=True),` in the dataloader makes data loading extremely slow.

The setting is that height_buckets = [480], width_buckets=[720] and frame_buckets=[49]. If using the BucketSampler, on 8 GPU processes, it will take 10 minutes to load the first batch of data. However, without using the BucketSampler, it only takes second.

I did some digging and found that even when batch_size=1 (with just one process), whenever we do next(iter(dataloader)),  in "\__iter__\" function of the BuckSampler [here](https://github.com/a-r-r-o-w/finetrainers/blob/4b21d37c9d4ae937880ef90f1a2b3d00fc2c832e/training/cogvideox/dataset.py#L405), the `for index, data in enumerate(self.data_source):` will actually loop more than once (even though batch_size=1 and we only have one bucket). I haven't figured out why. Any thoughts on this? Thank you1

**Comment by a-r-r-o-w** [comment]
Hey, thanks for your interest! Also nice to see you here again :)

The current dataset implementation is very bad tbh. It needs a complete rewrite, which I will try to get out in the next few days.

I don't quite see a 10 minute delay, but it's definitely much slower than one would expect it to be. Will prioritize this as soon as I find some free time. Would recommend using `--precompute_conditions` until then so that the data has to be preprocessed only once, and you can reuse it across multiple runs.

**Comment by Yuancheng-Xu** [comment]
Hi @a-r-r-o-w Thanks for the consistent efforts! It is nice to see this repo gradually supports training all kinds of models!

It seems that `--precompute_conditions` only saves time for VAE and text embedding computation. However, even with `--precompute_conditions` I notice significant overhead in data loading. To replicate, we can compare the current version with the one with `sampler=BucketSampler(train_dataset, batch_size=args.train_batch_size, shuffle=True),` commented out [here](https://github.com/a-r-r-o-w/finetrainers/blob/4b21d37c9d4ae937880ef90f1a2b3d00fc2c832e/training/cogvideox/cogvideox_text_to_video_lora.py#L522). It might also help to increase the datasize (maybe just create some dummy data to increase data size) to see the significant overhead. 

Does anyone else notice such overhead or it is just me?

**Comment by cdfan0627** [comment]
Have you tried the official [BucketSampler?](https://github.com/THUDM/CogVideo/blob/CogVideoX_dev/finetune/datasets/i2v_dataset.py#L276) Do you encounter the same issue?

**Comment by Yuancheng-Xu** [comment]
@cdfan0627 In my experience, yes.
